### PR TITLE
Select time-series indices in METRICS command

### PR DESCRIPTION
**Pull request by dnhatn**
With this change, the METRICS command will select only time-series indices.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java [hunk 1]
```diff
@@ -265,14 +271,17 @@ private EsField conflictingMetricTypes(String name, String fullName, FieldCapabi
...
+    private static FieldCapabilitiesRequest createFieldCapsRequest(String index, QueryBuilder indexFilter, Set<String> fieldNames) {
         FieldCapabilitiesRequest req = new FieldCapabilitiesRequest().indices(Strings.commaDelimitedListToStringArray(index));
         req.fields(fieldNames.toArray(String[]::new));
         req.includeUnmapped(true);
         // lenient because we throw our own errors looking at the response e.g. if something was not resolved
         // also because this way security doesn't throw authorization exceptions but rather honors ignore_unavailable
         req.indicesOptions(FIELD_CAPS_INDICES_OPTIONS);
         req.setMergeResults(false);
+        if (indexFilter != null) {
+            req.indexFilter(indexFilter);
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:283]: There is a reason why we didn't include the index filter in the field_caps request. A query needs to be syntactically correct, irrespective of whatever filters exist in the body of the esql request. I know that the filter on a field_caps request can **only** exclude entire indices, but it is doing that by looking at the data of those indices. Filtering data (and to some degree, indices) shouldn't interfere with the "metadata" of the indices (meaning, the mappings). The mappings should remain unchanged/fixed, no matter how the data is filtered.

Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:283]: I can pass the index mode in the field-caps request, but that’s unnecessarily complicated. This approach is more streamlined. Unlike a data filter, this index filter doesn’t change mappings or metadata based on the data in the target indices; it simply excludes non-time-series indices. I can make the field-caps implementation aware of this filter too, but that’s just an optimization.

Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:283]: When we assumed that a `metrics` source command implicitly considers the command `where _index_mode == "metricsdb"` I also assumed that we will get the fields for the index pattern specified in the `metrics` command but filter the rows based on `_index_mode`.

There is another implication of filtering the `field_caps` request: doing it means that we are likely filtering **columns (fields)** for the specified index pattern, doing it with `_index_mode == "metricsdb"` it means we are filtering the rows (the columns remain unchanged and they will cover the index pattern specified).

For example, `test1` is a tsdb index, `test2` is a regular index.
```
test1
-----
@timestamp
counter1
counter2
```

```
test2
-----
first_name
last_name
```

`metrics test*` will have the following columns in the result:

```
@timestamp   |   counter1   |   counter2  |   first_name   |   last_name
```

if we would to filter _field_caps with `index_filter` the columns will be:

```
@timestamp   |   counter1   |   counter2
```

Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:283]: Yes, that's expected. The first_name and last_name fields from regular indices should be excluded from the result.

Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:283]: The `metrics test*` command should target only time-series indices that start with `test`.

Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:283]: Imho, this is not correct.
Columns should be filtered out in the query, if the users wants to, with `drop` and `keep` commands. `_index_mode == "metricsdb"` should filter only the rows.

Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:283]: It was a rough equivalent we tried to capture, but it wasn’t precisely equivalent. The metrics command should exclude non-time-series indices, including their metadata.

**Comment by astefan** [comment]
After an offline discussion, we believe it is acceptable for the `metrics` command to pre-filter indices in the `field_caps` request, in the absence of a catalog-like service in Elasticsearch.

Imo, we should restrict the IndexResolver usage of this `metrics` specific index filter as much as possible by:
- `resolveAsMergedMapping` shouldn't receive a generic QueryBuilder as index filter, but maybe only a `boolean` value indicating the request comes from `metrics` or not
- the `metrics` specific filter should have a `static final` instance and be used only in `IndexResolver`
- the `TableIdentifier` should use the above mentioned `boolean` instead of the generic `QueryBuilder`.

**Comment by dnhatn** [comment]
Thanks, Andrei. I’m closing this PR and incorporating the changes here into https://github.com/elastic/elasticsearch/pull/112383.

**Pull request closed without merging**